s3transfer/download.py

```diff
@@ -0,0 +1,311 @@
+from s3transfer.utils import get_callbacks
+from s3transfer.utils import invoke_progress_callbacks
+from s3transfer.utils import StreamReaderProgress
+from s3transfer.tasks import Task
+from s3transfer.tasks import TaskSubmitter
+
+
+logger = logging.getLogger(__name__)
+
+S3_RETRYABLE_ERRORS = (
+    socket.timeout, SOCKET_ERROR, ReadTimeoutError, IncompleteReadError
+)
+
+
+class DownloadTaskSubmitter(TaskSubmitter):
+    def __init__(self, client, config, osutil, executor, io_executor):
+        super(DownloadTaskSubmitter, self).__init__(
+            client, config, osutil, executor)
+        self._io_executor = io_executor
+
+    def _submit(self, transfer_future, transfer_coordinator):
+        if transfer_future.meta.size is None:
+            # If a size was not provided figure out the size for the
+            # user.
+            response = self._client.head_object(
+                Bucket=transfer_future.meta.call_args.bucket,
+                Key=transfer_future.meta.call_args.key,
+                **transfer_future.meta.call_args.extra_args
+            )
+            transfer_future.meta.provide_transfer_size(
+                response['ContentLength'])
+
+        # Figure out a temporary filename for the file.
+        temp_filename = (
+            transfer_future.meta.call_args.fileobj + os.extsep +
+            random_file_extension()
+        )
+        # If it is greater than threshold do a ranged download, otherwise
+        # do a regular GetObject download.
+        if transfer_future.meta.size < self._config.multipart_threshold:
+            self._submit_download_request(
+                temp_filename, transfer_future, transfer_coordinator)
+        else:
+            self._submit_ranged_download_request(
+                temp_filename, transfer_future, transfer_coordinator)
+
+    def _submit_download_request(
+            self, temp_filename, transfer_future, transfer_coordinator):
+        call_args = transfer_future.meta.call_args
+
+        # Get a handle to the temp file
+        temp_f = self._get_io_file_handle(temp_filename, transfer_coordinator)
+
+        # Get the needed callbacks for the task
+        progress_callbacks = get_callbacks(transfer_future, 'progress')
+        done_callbacks = get_callbacks(transfer_future, 'done')
+
+        # Before starting any tasks for downloads set up a cleanup function
+        # that will make sure that the temporary file always gets
+        # cleaned up.
+        transfer_coordinator.add_failure_cleanup(
+            self._osutil.remove_file, temp_filename)
+
+        # Submit the task to download the object.
+        download_future = self._executor.submit(
+            GetObjectTask(
+                transfer_coordinator=transfer_coordinator,
+                main_kwargs={
+                    'client': self._client,
+                    'bucket': call_args.bucket,
+                    'key': call_args.key,
+                    'f': temp_f,
```

Comment:
Can we use something more description than `f`.  I'm ok with the convention of `f` for a fileobj in a local scope, but if we're passing it around as kwargs to other objects I prefer a more descriptive name like `fileobj`, etc.